numpy/lib/function_base.py

```diff
@@ -3670,17 +3883,90 @@ def _quantile_is_valid(q):
     return True
 
 
-def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
-                           interpolation='linear', keepdims=False):
+def _validate_weights(a, q, axis, weights):
+    if weights is None:
+        return None
+    else:
+        a = asanyarray(a)
+        wgt = np.asanyarray(weights)
+
+        if not np.issubdtype(wgt.dtype, np.number):
+            # convert every element to np.float
+            # raises ValueError if any element fails to convert
+            wgt = np.vectorize(np.float)(wgt)
+
+        if np.isnan(wgt).any():
+            raise ValueError("No weight can be NaN.")
+
+        if (wgt < 0).any():
+            raise ValueError("Negative weight not allowed.")
+
+        if issubclass(a.dtype.type, np.bool_):
+            raise TypeError("Boolean weights not supported")
+        elif issubclass(a.dtype.type, np.integer):
+            result_dtype = np.result_type(a.dtype, wgt.dtype, 'f8')
+        else:
+            result_dtype = np.result_type(a.dtype, wgt.dtype)
+
+        # determine if the weights can be applied to the array
+        if a.shape == wgt.shape or wgt.size == 1:
+            broadcastable = True
+        elif a.ndim == wgt.ndim:  # same ndim, but some dims can be size = 1
+            broadcastable = all([a_dim == w_dim or w_dim == 1
+                                 for a_dim, w_dim
+                                 in zip(a.shape, wgt.shape)])
+        else:
+            broadcastable = False
+
+        if not broadcastable:  # 1-D array needs preprocessing before broadcast
+            if axis is None:
+                raise TypeError(
+                    "Axis must be specified when shapes of a and weights "
+                    "differ and not broadcastable.")
+            if wgt.ndim != 1:
+                raise TypeError(
+                    "1D weights expected when shapes of a and weights differ "
+                    " and not broadcastable.")
+            if wgt.shape[0] != a.shape[axis]:
+                raise ValueError(
+                    "Length of weights not compatible with specified axis.")
+            # set up wgt to broadcast along axis
+            wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+            wgt = wgt.swapaxes(-1, axis)
+        else:  # same shape, or at least broadcastable
+            if axis is None:
+                # to avoid ap = a.flatten() in _quantile_ureduce_func,
+                # we explicitly set axis
+                axis = tuple(range(a.ndim))
+
+        scl = wgt.sum(axis=axis, dtype=result_dtype)
+        if np.any(scl == 0.0):
+            raise ZeroDivisionError(
+                "Weights sum to zero, can't be normalized")
+
+        # Obtain a weights array of the same shape as reduced a
+        wgt = np.broadcast_to(wgt, a.shape)
```

Comment:
This can move inside the "or at least broadcastable" branch above

Reply:
I think it'll be better to leave this one outside.  In the `not broadcastable` branch, I prep the wgt array so that it would be broadcastable.  so `wgt = np.broadcast_to(wgt, a.shape)` would be called regardless of which code path it took above.